File: frontend/src/lib/api/voices.ts
import axiosInstance from '$lib/axios';

export type ReferenceTextStatus = 'pending' | 'ready' | 'failed' | 'manual';

export interface VoiceProfile {
  id: string;
  name: string;
  model_id: string;
  reference_audio_path: string;
  embedding_path: string | null;
  metadata?: Record<string, unknown> | null;
  extra_info?: Record<string, unknown> | null;
  description?: string | null;
  tags?: string[];
  reference_text: string | null;
  reference_text_status: ReferenceTextStatus;
  reference_language: string | null;
  created_at: string;
  /** Mean speaker_similarity across completed jobs that used this profile.
   *  ``null`` when no scored jobs exist yet. */
  avg_similarity: number | null;
  /** Number of scored completed jobs contributing to ``avg_similarity``. */
  similarity_count: number;
  /** Optional duration of the reference audio in seconds. Backend may omit. */
  reference_duration_seconds?: number | null;
}

export interface VoiceProfileUpdate {
  name?: string;
  description?: string;
  tags?: string[];
  reference_text?: string | null;
  reference_language?: string | null;
}

export interface VoiceListResponse {
  voices: VoiceProfile[];
  total: number;
}

export interface BuiltinVoice {
  id: string;
  name: string;
  language: string;
  gender: string | null;
  model_id: string;
}

export async function listVoices(modelId?: string): Promise<VoiceListResponse> {
  const res = await axiosInstance.get<VoiceListResponse>('/voices', {
    params: modelId ? { model_id: modelId } : undefined,
  });
  return res.data;
}

export async function getVoice(voiceId: string): Promise<VoiceProfile> {
  const { data } = await axiosInstance.get<VoiceProfile>(`/voices/${voiceId}`);
  return data;
}

export async function createVoiceProfile(
  name: string,
  modelId: string,
  referenceAudio: File,
  referenceText?: string
): Promise<VoiceProfile> {
  const form = new FormData();
  form.append('name', name);
  form.append('model_id', modelId);
  form.append('reference_audio', referenceAudio);
  if (referenceText !== undefined && referenceText !== null && referenceText !== '') {
    form.append('reference_text', referenceText);
  }
  const res = await axiosInstance.post<VoiceProfile>('/voices', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return res.data;
}

export async function updateVoice(
  voiceId: string,
  update: VoiceProfileUpdate
): Promise<VoiceProfile> {
  const { data } = await axiosInstance.patch<VoiceProfile>(`/voices/${voiceId}`, update);
  return data;
}

/** Convenience wrapper: PATCH the transcript only. Empty string clears it
 *  (server resets status to "pending" and re-runs ASR). */
export async function updateVoiceTranscript(
  voiceId: string,
  text: string
): Promise<VoiceProfile> {
  return updateVoice(voiceId, { reference_text: text });
}

/** Trigger a fresh ASR run for the profile. Server resets status to "pending". */
export async function transcribeVoice(voiceId: string): Promise<VoiceProfile> {
  const { data } = await axiosInstance.post<VoiceProfile>(`/voices/${voiceId}/transcribe`);
  return data;
}

export async function deleteVoiceProfile(voiceId: string): Promise<void> {
  await axiosInstance.delete(`/voices/${voiceId}`);
}

export async function listBuiltinVoices(modelId: string): Promise<BuiltinVoice[]> {
  const res = await axiosInstance.get<BuiltinVoice[]>(`/voices/builtin/${modelId}`);
  return res.data;
}

export function getVoiceAudioUrl(voiceId: string): string {
  return `/api/voices/${voiceId}/audio`;
}

File: backend/app/schemas/voices.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

# Status values for the reference-transcript pipeline.
ReferenceTextStatus = Literal["pending", "ready", "failed", "manual"]

# Maximum length of a reference transcript (auto-detected or manually entered).
# Anything beyond this is almost certainly the wrong field; the upstream
# upload handler enforces the same cap with a 422 response.
MAX_REFERENCE_TEXT_LEN = 4000


def _normalise_reference_text(value: str | None) -> str | None:
    """Strip surrounding whitespace, reject control chars (except \\n / \\t),
    and treat empty strings as ``None`` so downstream ``if profile.reference_text``
    checks remain uniform."""
    if value is None:
        return None
    cleaned = value.strip()
    if not cleaned:
        return None
    for ch in cleaned:
        if ord(ch) < 0x20 and ch not in ("\n", "\t"):
            raise ValueError("reference_text contains disallowed control characters")
    return cleaned


class VoiceProfileCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    model_id: str = Field(..., description="Which model this voice is for")
    reference_text: str | None = Field(default=None, max_length=MAX_REFERENCE_TEXT_LEN)

    @field_validator("reference_text", mode="before")
    @classmethod
    def _validate_reference_text(cls, v: str | None) -> str | None:
        return _normalise_reference_text(v)


class VoiceProfileResponse(BaseModel):
    id: uuid.UUID
    name: str
    model_id: str
    reference_audio_path: str
    embedding_path: str | None
    extra_info: dict | None
    description: str | None = None
    tags: list = Field(default_factory=list)
    reference_text: str | None = None
    reference_text_status: ReferenceTextStatus = "pending"
    reference_language: str | None = None
    created_at: datetime
    avg_similarity: float | None = Field(
        None,
        description=(
            "Mean speaker_similarity across all completed jobs that used this voice "
            "profile. NULL when no scored jobs exist yet."
        ),
    )
    similarity_count: int = Field(
        0,
        description=(
            "Number of completed jobs with a non-null speaker_similarity for this profile."
        ),
    )

    model_config = {"from_attributes": True}


class VoiceProfileUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=128)
    description: str | None = None
    tags: list[str] | None = None
    reference_text: str | None = Field(default=None, max_length=MAX_REFERENCE_TEXT_LEN)
    reference_language: str | None = Field(default=None, max_length=8)

    @field_validator("reference_text", mode="before")
    @classmethod
    def _validate_reference_text(cls, v: str | None) -> str | None:
        return _normalise_reference_text(v)


class VoiceListResponse(BaseModel):
    voices: list[VoiceProfileResponse]
    total: int


class BuiltinVoice(BaseModel):
    """A built-in (non-cloned) voice shipped with a model."""

    id: str
    name: str
    language: str
    gender: str | None = None
    model_id: str


class SimilarityTestResponse(BaseModel):
    """Response from POST /api/voices/{id}/test."""

    similarity: float
